notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+    "COL_TIMESTAMP = \"Timestamp\"\n",
+    "COL_RATING = \"Rating\"\n",
+    "COL_PREDICTION = \"prediction\"\n",
+    "\n",
+    "HEADER = {\n",
+    "    \"col_user\": COL_USER,\n",
+    "    \"col_item\": COL_ITEM,\n",
+    "    \"col_rating\": COL_RATING,\n",
+    "    \"col_prediction\": COL_PREDICTION,\n",
+    "}\n",
+    "\n",
+    "HEADER_ALS = {\n",
+    "    \"userCol\": COL_USER,\n",
+    "    \"itemCol\": COL_ITEM,\n",
+    "    \"ratingCol\": COL_RATING\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1 Data preparation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For illustration purpose, data used for demonstrating the hyperparameter tuning practices is the Movielens 100k dataset. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "spark = start_or_get_spark()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Movielens 100k dataset is used for running the demonstration."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data = load_spark_df(spark, size='100k')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataset is split into 3 subsets randomly with a given split ratio. The hyperparameter tuning is performed on the training and the validating data, and then the optimal recommender selected is evaluated on the testing dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train, valid, test = spark_random_split(data, ratio=[4, 4, 2])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2 Hyper parameter tuning with Azure Machine Learning Services"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `hyperdrive` module in the [Azure Machine Learning Services](https://azure.microsoft.com/en-us/services/machine-learning-service/) is supposed to run [hyperparameter tuning and optimizing for machine learning model selection](https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-tune-hyperparameters). At the moment, the service supports running hyperparameter tuning on heterogenous computing targets such as cluster of commodity compute nodes with or without GPU devices (see detailed documentation [here](https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-set-up-training-targets)). It is feasible to run parameter tuning on a cluster of VM nodes. In this case, the service containerizes individual and independent Spark session on each node of the cluster to run the parameter tuning job in parallel, instead of inside a single Spark session where the training is executed in a distributed manner.  \n",
```

Comment:
This section is weird, we are talking about a library that we are not using and that will be available in the future (but when?). I would vote to remove it or push it to the bottom after the other 2 hyperparameter methods